File: src/core/tools/tool.ts
import type { Canvas } from "../canvas";
import type { Vector } from "../vector";

export interface Mods {
  ctrl: boolean;
  alt: boolean;
  /** Reverse elbow orientation: Ctrl, or the `f` toggle (§6). */
  flip: boolean;
}

export const NO_MODS: Mods = { ctrl: false, alt: false, flip: false };

export function mods(partial: Partial<Mods> = {}): Mods {
  const m = { ...NO_MODS, ...partial };
  return { ...m, flip: partial.flip ?? m.ctrl };
}

/** Non-printable keys, as ASCIIFlow names them. */
export const KEY = {
  ENTER: "<enter>",
  BACKSPACE: "<backspace>",
  DELETE: "<delete>",
  UP: "<up>",
  DOWN: "<down>",
  LEFT: "<left>",
  RIGHT: "<right>",
} as const;

/** Terminal stand-in for the CSS cursor. */
export type HoverHint = "default" | "crosshair" | "move" | "resize-h" | "resize-v" | "text";

export interface ToolContext {
  readonly canvas: Canvas;
}

export interface Tool {
  start(p: Vector, m: Mods): void;
  move(p: Vector, m: Mods): void;
  end(): void;
  /** Tool switched away. */
  cleanup(): void;
  /** Returns true when the key was consumed. */
  handleKey(key: string, m: Mods): boolean;
  hoverHint(p: Vector, m: Mods): HoverHint;
}

File: src/core/tools/tool.rs
//! Shared tool vocabulary: modifiers, keys, hover hints.
//!
//! Ported from `ASCIIFlow`'s tool contract; rsdia's tools take the canvas as an
//! explicit parameter instead of holding a reference back to the editor.

use crate::core::canvas::Canvas;
use crate::core::vector::Pos;

#[derive(Clone, Copy, Debug, PartialEq, Eq, Default)]
pub struct Mods {
    pub ctrl: bool,
    pub alt: bool,
    /// Reverse elbow orientation: Ctrl, or the `f` toggle.
    pub flip: bool,
}

impl Mods {
    pub const NONE: Self = Self {
        ctrl: false,
        alt: false,
        flip: false,
    };

    /// Ctrl also flips, matching `ASCIIFlow`'s shift-drag.
    #[must_use]
    pub const fn new(ctrl: bool, alt: bool) -> Self {
        Self {
            ctrl,
            alt,
            flip: ctrl,
        }
    }

    #[must_use]
    pub const fn flipped(mut self, flip: bool) -> Self {
        self.flip = flip;
        self
    }
}

/// A tool-facing key: a printable character or one of the named keys.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Key {
    Char(char),
    Enter,
    Backspace,
    Delete,
    Up,
    Down,
    Left,
    Right,
}

/// Tools receive the canvas for each call; they never store a reference to it.
pub trait Tool {
    fn start(&mut self, canvas: &mut Canvas, p: Pos, m: Mods);
    fn move_to(&mut self, canvas: &mut Canvas, p: Pos, m: Mods);
    fn end(&mut self, canvas: &mut Canvas);
    /// Tool switched away.
    fn cleanup(&mut self, canvas: &mut Canvas);
    /// Returns true when the key was consumed.
    fn handle_key(&mut self, canvas: &mut Canvas, key: Key, m: Mods) -> bool;
    /// Whether the cell under the pointer is something this tool would grab.
    fn hover_is_target(&self, canvas: &Canvas, p: Pos, m: Mods) -> bool;
}
